b2/transferer.py

```diff
@@ -167,14 +167,21 @@ def _download_file_parallel(
         bytes_read = file.tell()
         return bytes_read, hasher.hexdigest()
 
-    def _get_parts(self, response, writer, hasher, part_size, number_of_streams):
-        stream = FirstPartDownloaderThread(response, self.session, writer, hasher, (0, part_size))
+    def _get_parts(self, response, writer, hasher, ideal_part_size, number_of_streams):
+        end_of_first_chunk = int(ideal_part_size)
+        stream = FirstPartDownloaderThread(
+            response,
+            self.session,
+            writer,
+            hasher,
+            (0, end_of_first_chunk),
+        )
         stream.start()
         streams = [stream]
 
-        start = part_size
+        start = end_of_first_chunk
         for part_number in range(1, number_of_streams):
-            end = (part_number + 1) * part_size
+            end = int((part_number + 1) * ideal_part_size)
```

Comment:
Floating point numbers make me nervous.  Roundoff error could cause this truncation to be one less than expected.  On the other hand, I think there's an off-by-one error here because the end of a range is _inclusive_, which will try to make the end of the last part one higher than it should be.  When it downloads, though, HTTP says the server will clip the range to what's available in the file.

Reply:
This needs to be fixed, the user might want to download a part of a file, then we'd retrieve one byte too much.

Do you have a suggestion on how to implement this properly?

Reply:
I like math puzzles like this, so I was playing with it last night.  This is what I came up with:

```
    """
    HTTP ranges use an *inclusive* index at the end.
    """

    def __init__(self, start, end):
        assert 0 <= start <= end
        self.start = start
        self.end = end

    def size(self):
        return self.end - self.start + 1

    def subrange(self, sub_start, sub_end):
        """
        Returns a range that is part of this range.
        :param sub_start: Index relative to the start of this range.
        :param sub_end: (Inclusive!) index relative to the start of this range.
        :return: A new Range
        """
        assert 0 <= sub_start <= sub_end < self.size()
        return Range(self.start + sub_start, self.start + sub_end)

    def __repr__(self):
        return 'Range(%d, %d)' % (self.start, self.end)


class PartToDownload(object):
    """
    Holds the range of a file to download, and the range of the
    local file where it should be stored.
    """
    def __init__(self, cloud_range, local_range):
        self.cloud_range = cloud_range
        self.local_range = local_range

    def __repr__(self):
        return 'PartToDownload(%s, %s)' % (self.cloud_range, self.local_range)


def gen_parts(cloud_range, local_range, part_count):
    """
    Generates a sequence of PartToDownload to download a large file as
    a collection of parts.
    """
    assert cloud_range.size() == local_range.size()
    assert 0 < part_count <= cloud_range.size()
    offset = 0
    remaining_size = cloud_range.size()
    for i in range(part_count):
        # This rounds down, so if the parts aren't all the same size,
        # the smaller parts will come first.
        this_part_size = remaining_size // (part_count - i)
        yield PartToDownload(
            cloud_range.subrange(offset, offset + this_part_size - 1),
            local_range.subrange(offset, offset + this_part_size - 1)
        )
        offset += this_part_size
        remaining_size -= this_part_size
```